crates/ruff_linter/src/rules/pycodestyle/rules/type_comparison.rs

```diff
@@ -37,90 +39,57 @@ pub struct TypeComparison;
     {
-        if !matches!(op, CmpOp::Is | CmpOp::IsNot | CmpOp::Eq | CmpOp::NotEq) {
-            continue;
+        if is_type(left, checker.semantic()) || is_type(right, checker.semantic()) {
+            checker
+                .diagnostics
+                .push(Diagnostic::new(TypeComparison, compare.range()));
         }
+    }
+}
 
-        // Left-hand side must be, e.g., `type(obj)`.
-        let Expr::Call(ast::ExprCall { func, .. }) = left else {
-            continue;
-        };
+/// Returns `true` if the [`Expr`] is known to evaluate to a type (e.g., `int`, or `type(1)`).
+fn is_type(expr: &Expr, semantic: &SemanticModel) -> bool {
+    match expr {
+        Expr::Call(ast::ExprCall {
+            func, arguments, ..
+        }) => {
+            // Ex) `type(obj) == type(1)`
+            let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
+                return false;
+            };
 
-        let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
-            continue;
-        };
+            if !(id == "type" && semantic.is_builtin("type")) {
+                return false;
+            };
 
-        if !(id == "type" && checker.semantic().is_builtin("type")) {
-            continue;
+            // Allow comparison for types which are not obvious.
+            arguments
+                .args
+                .first()
+                .is_some_and(|arg| !arg.is_name_expr() && !is_const_none(arg))
         }
-
-        // Right-hand side must be, e.g., `type(1)` or `int`.
-        match right {
-            Expr::Call(ast::ExprCall {
-                func, arguments, ..
-            }) => {
-                // Ex) `type(obj) is type(1)`
-                let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
-                    continue;
-                };
-
-                if id == "type" && checker.semantic().is_builtin("type") {
-                    // Allow comparison for types which are not obvious.
-                    if arguments
-                        .args
-                        .first()
-                        .is_some_and(|arg| !arg.is_name_expr() && !is_const_none(arg))
-                    {
-                        checker
-                            .diagnostics
-                            .push(Diagnostic::new(TypeComparison, compare.range()));
-                    }
-                }
-            }
-            Expr::Attribute(ast::ExprAttribute { value, .. }) => {
-                // Ex) `type(obj) is types.NoneType`
-                if checker
-                    .semantic()
-                    .resolve_call_path(value.as_ref())
-                    .is_some_and(|call_path| matches!(call_path.as_slice(), ["types", ..]))
-                {
-                    checker
-                        .diagnostics
-                        .push(Diagnostic::new(TypeComparison, compare.range()));
-                }
-            }
```

Comment:
This isn't present in the refactored code, is that an expected behavior?

The following doesn't get violated then:
```python
if type(foo) == types.LambdaType:
    ...
```

Reply:
Yeah, they removed this from pycodestyle with the understanding that a user-defined `types` would be a lot more common than accessing the standard library's `types`. I'm torn on it but figured it's fine to follow suit.